elytron/WFLY-14735-Management-FORM-authentication.adoc

```diff
@@ -0,0 +1,245 @@
+
+Updates will be required to the https://github.com/hal/console[Admin Console] to support
+possibly being required to redirect to the login form and to make use of the logout endpoint.
+
+Security components developed for this RFE will be contained within the
+https://github.com/wildfly-security/wildfly-elytron[WildFly Elytron] project.
+
+=== Other Interested Projects
+
+None
+
+=== Relevant Installation Types
+// Remove the x next to the relevant field if the feature in question is not relevant
+// to that kind of WildFly installation
+* [x] Traditional standalone server (unzipped or provisioned by Galleon)
+
+* [x] Managed domain
+
+* [x] OpenShift s2i
+
+* [x] Bootable jar
+
+== Requirements
+
+=== Hard Requirements
+
+==== General Flow
+
+On the HTTP management interface a new context will be made available at the path `/authentication`,
+when a user connects to the root `/` context of the application server they will be redirected to the
+`/authentication` context.  The `/authentication` context will display a login form for the user 
+to authenticate themselves.
+
+The look and feel of the login FORM will make use of `PatternFly` for consistent with the look and
+feel of the web console.
+
+Under the `/authentication` context there will be two sub contexts `/login` and `/logout`, after 
+a user has input their username and password these will be passed to the first of the sub contexts to
+perform the authentication.  Where authentication is successful a cryptographically signed JWT token 
+will be generated and passed back to the client as a cookie, the login form will then redirect the 
+user to the admin console.
+
+The admin console context is unauthenticated so will load as normal, the `/management` enpoint will 
+use a different HTTP authentication mchanism configuration which will use the `BEARER` HTTP 
+authentication mechanism which will be configured to use the JWT token from the incoming request cookie.
+
+When the user has finished managing the application server they can click the usual Logout link in 
+the admin console which will redirect them to `/authentication/logout`, this context will clear 
+the cookie containing the JWT token.
+
+==== Other Flows
+
+If after the username and password is sent to the `/authentication/login` context the authentication 
+fails the user will remain on the login page and a suitably vague error will be delayed to indicate
+that authentication has failed.
+
+If a user has previously been authenticated and their cookie has expired or the user is using the 
+admin console to connect to a different server it is possible the admin console may sent an invocation
+to the `/management` context without an appropriate cookie, in this case the admin console will need
+to be "notified to" and handle a redirect to the `/authentication` context.
+
+==== elytron-authentication
+
+A new subsystem `elytron-authentication` will be developed, the primary role of this subsystem
```

Comment:
Great question!

Reply:
As a new subsystem I expect this will be a distinct module, regarding code written outside of WildFly I haven't looked exactly at where it will be added but if possible it should be separate from our existing Elytron modules so can be provisioned independently.

This is a new piece of functionality that is optional i.e. current behaviour can be preserved - for this reason I think it is justified to define a new layer which will include this new subsystem/extension and provide a default definition within the subsystem that depends on the ManagementDomain.  This will however not make it in use by default.

Maybe a follow on discussion for elsewhere, would layers be suitable granularity for some chunks of default security configuration - i.e. break it down further.  A layer for application SSL, one for management SSL, one for existing management domain config, one for new config.  Could that instead be a use for an external feature pack to define chunks of security config which can be selected?

Reply:
I expect the module to be private as we don't need to expose any new API for things like custom realms.

Do I understand it well there is no mechanism to customize layers and we need a layer per specific case? If it's the case is there a recommendation for naming schemas?

Reply:
+1 as an extension / subsystem it will be a private module.

For the layer under this RFE I think the name should be very close to the name of the subsystem.